apps/prairielearn/src/lib/sentry.ts

```diff
@@ -0,0 +1,72 @@
+import * as Sentry from '@prairielearn/sentry';
+import type { NextFunction, Request, Response } from 'express';
+
+export function enrichSentryEventMiddleware(req: Request, res: Response, next: NextFunction) {
+  // This will ensure that this middleware is always run in an isolated
+  // context so that we don't accidentally leak data between requests.
+  Sentry.runWithAsyncContext(() => {
+    const hub = Sentry.getCurrentHub();
+
+    hub.configureScope((scope) => {
+      // This event processor will run whenever an event is captured by Sentry.
+      // It will use as much context as it can, but will gracefully handle the
+      // case where context is missing.
+      scope.addEventProcessor((event) => {
+        event.tags ??= {};
+
+        event.tags.response_id = res.locals.response_id;
+        event.tags.method = req.method;
+        event.tags.url = req.originalUrl;
+
+        if (res.locals.error_id) {
+          event.tags.error_id = res.locals.error_id;
+        }
+
+        if (res.locals.authn_user?.user_id) {
+          event.user = {
+            id: res.locals.authn_user.user_id.toString(),
+            // To comply with GDPR and other data protection laws, you can
+            // configure Sentry to not store IP addresses. Sentry will then
+            // only use the IP address to compute a country code and
+            // immediately discard it.
+            ip_address: req.ip,
+          };
+        }
+
+        if (res.locals?.course?.id) {
+          event.tags['course.id'] = res.locals.course?.id;
+          event.tags['course.short_name'] = res.locals.course?.short_name;
+          event.tags['course.title'] = res.locals.course?.title;
+        }
```

Comment:
There is exactly one reason that I like dumping request context onto `res.locals`, and it's the fact that we can do things like this without spreading calls to `Sentry.configureScope` across each and ever one of our route handlers 😅

Matt and I were talking a bit the other day about moving absolutely everything out of `res.locals` (at least in PT), and this is a case where we should think carefully about if we actually want to do this, and if we do, whether we can maintain this useful property of `res.locals` some other way.

Reply:
Yep, this is definitely a good point. My thinking here is that in every route handler we'd need to do something like:
```
router.get(... {
    const agentUser = await performAuthn(req);
    const course = await selectAndAuthzCourse({unsafe_course_id: req.params.unsafe_course_id, agentUser});
    await enhancedLog(req, {agentUser, course});
    ...
});
```
Here the `enhanceLog()` call would take any extra context that the page might have at that point. This would be used to actually write an improved log message, to store extra data in `res.locals` for the response logging middleware, and to enrich the sentry scope.

This is clearly uglier than having everything automagically handled by middleware, but I think the improvement in having all of our authorization code strongly typed (and having a clear call sequence) is worth it. More of the extra code is really coming from calling `selectAndAuthzCourse()` in every route, rather than having a single middleware do this authz and store the result in `res.locals`, but I already think that it's so much clearer having it in the route handler that I don't care about the extra code duplication.